### PR TITLE
feat(plugins): Expose `ConstraintEvaluator` as an extension point

### PR DESCRIPTION
**Pull request by robzienert**
Moves all of the core constraints code into `keel-api`.

Two new interfaces were added to help this effort:

- `com.netflix.spinnaker.keel.api.support.EventPublisher` - A subset interface
  of Spring's `ApplicationEventPublisher`. This is injected into `ConstraintEvaluator`
  classes as `SpringEventPublishedBridge`, which just delegates to Spring.
- `com.netflix.spinnaker.keel.api.constraints.ConstraintRepository` - This is
  just a subset of the god `KeelRepository` interface. Implemented by
  `ConstraintRepositoryBridge`, which just delegates to the `KeelRepository`
  implementation.

This also introduces the `ULID` library as an api dependency of `keel-api`.

I also did some non-functional cleanup: Single class per file rule.

⚠️ I didn't bother documenting all of the interfaces introduced to `keel-api` that already existed - it looks like the MD team hasn't really been doing many (any?) code docs -- I'm leaving this as an exercise for them to take on. If external contributions to Keel are desired, having docs on everything in `keel-api` is going to be needed rather than relying on tribal knowledge of contracts and behaviors.

### CODE REVIEW

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/constraints/ConstraintEvaluator.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.support.EventPublisher
+import org.pf4j.ExtensionPoint
+
+/**
+ * TODO: Docs
+ */
+interface ConstraintEvaluator<T : Constraint> : ExtensionPoint {
```
Comment (jonsie) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/constraints/ConstraintEvaluator.kt:28]: Use `SpinnakerExtensionPoint` instead.

**Comment by lorin** [comment]
@robzienert Is it possible to split this up into smaller PRs? Otherwise, I fear it'll take me much longer to review this, and the increased scope and time increases the chances of having to deal with merge conflicts.

**Comment by robzienert** [comment]
@lorin I can try. Almost all of this code is just moving files from one place to another, so there's not really much going on here. I'll get around to this sometime in the next couple weeks.

**Comment by lorin** [comment]
@robzienert Given that it's mostly just moving files around, if you just resolve the current merge conflicts, then I can power through a review of the PR as-is.

**Comment by luispollo** [comment]
> @robzienert Given that it's mostly just moving files around, if you just resolve the current merge conflicts, then I can power through a review of the PR as-is.

Same.

**Comment by robzienert** [comment]
Updated. Will look to do smaller PRs in the future.

**Comment by luispollo** [comment]
@robzienert I'm reviewing this, and appreciate the contribution and the details in the PR description, but the context I missed was on the _why_: what is the _goal_ of the change?

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/constraints/ConstraintEvaluator.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+interface ConstraintEvaluator<T : Constraint> : SpinnakerExtensionPoint {
+
+  companion object {
+    /**
+     * TODO: Docs
+     */
+    fun <T> getConstraintForEnvironment(
+      deliveryConfig: DeliveryConfig,
+      targetEnvironment: String,
+      klass: Class<T>
```
Comment (luispollo) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/constraints/ConstraintEvaluator.kt:37]: Could you please take this opportunity to rename this to `constraintType` for clarity?

**Comment by robzienert** [comment]
> P.S. As far as I can tell, SpinnakerExtensionPoint is just a marker interface. Could you please point to code or docs that explain its intent and usage?

It is primarily a marker interface, explicitly calling out what is and is not an extension point. It is also used within the plugin framework for handling some proxy behavior (we proxy extension points for metrics, logging, etc) which we need a common root for.

https://github.com/spinnaker/kork/blob/master/kork-plugins-api/src/main/java/com/netflix/spinnaker/kork/plugins/api/internal/SpinnakerExtensionPoint.java

> Also, should the other plugin interfaces (ResourceHandler, Resolver) implement the interface as well?

They should, yes. But they first must be moved to `keel-api`.

**Pull request merged**